File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        ArrayList<HashSet<Integer>> arr = new ArrayList<>();
        int n = N;
        int answer = -1;

        for (int i = 1; i < 9; i++) {
            HashSet<Integer> set = new HashSet<>();

            set.add(n);
            n = n * 10 + N;

            for (int j = 0; j < i - 1; j++) {
                for (int op1 : arr.get(j)) {
                    for (int op2 : arr.get(arr.size() - j - 1)) {
                        set.add(op1 + op2);
                        set.add(op1 - op2);
                        set.add(op1 * op2);

                        if (op2 != 0) {
                            set.add(op1 / op2);
                        }
                    }
                }
            }

            if (set.contains(number)) {
                answer = i;
                break;
            }

            arr.add(set);
        }

        return answer;
    }
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.kt
class Solution {
    fun solution(N: Int, number: Int): Int {
        val dp = arrayListOf<MutableSet<Int>>()

        for (i in 1..8) {
            val case = mutableSetOf(N.toString().repeat(i).toInt())

            for (j in 0..(i - 2)) {
                for (op1 in dp[j]) {
                    for (op2 in dp[dp.size - j - 1]) {
                        case.add(op1 + op2)
                        case.add(op1 - op2)
                        case.add(op1 * op2)

                        if (op2 != 0) {
                            case.add(op1 / op2)
                        }
                    }
                }
            }

            if (case.contains(number)) {
                return i
            }

            dp.add(case)
        }

        return -1
    }
}